source/extensions/filters/network/connection_limit/connection_limit.cc

```diff
@@ -0,0 +1,100 @@
+#include "extensions/filters/network/connection_limit/connection_limit.h"
+
+#include "envoy/extensions/filters/network/connection_limit/v3/connection_limit.pb.h"
+
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ConnectionLimitFilter {
+
+Config::Config(
+    const envoy::extensions::filters::network::connection_limit::v3::ConnectionLimit& proto_config,
+    Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime)
+    : enabled_(proto_config.runtime_enabled(), runtime),
+      stats_(generateStats(proto_config.stat_prefix(), scope)),
+      max_connections_(PROTOBUF_GET_WRAPPED_REQUIRED(proto_config, max_connections)),
+      connections_(0), delay_(PROTOBUF_GET_OPTIONAL_MS(proto_config, delay)),
+      delay_timer_(delay_.has_value() && delay_.value() > std::chrono::milliseconds(0)
+                       ? dispatcher.createTimer([this] { onDelayTimer(); })
+                       : nullptr) {
+  if (delay_timer_) {
+    delay_timer_->enableTimer(delay_.value());
```

Comment:
Is it supposed to be a single timer for all connections?

Reply:
The timer here is only a test, I want to run a simple timer to verify, but unfortunately, it doesn't work for me, I never see the log which prints in the timer callback.
My aim is to use a timer to implement the delay closing connection, I need a timer enabled for every closing connection in the Filter class.
Can you help me with this? Many thanks!

Reply:
It might be that you run envoy with too low verbosity level. Make sure you add `-l trace` to its param list.

Reply:
Yes, I added the `-l trace`, am I using the timer correctly in this piece of code?
Should I enable to see the log duration of my integration tests? I am not sure about that.
Last question: Should I use this line `read_callbacks_->connection().dispatcher().createTimer()` to create a timer in my filter class if I want to implement the delay connection closing? 
Thanks!

Reply:
Yes, the timer code looks correct. Did you add  the filter to the listener's filter chain? I've just built and run envoy with your patch and I do see one "on delay timer" line in logs.

The line seems OK.

Reply:
Could you tell me about the steps of "built and run envoy with patch"? Sorry for not familiar with this very well.
I just ran my connection_limit_integration_test by using `bazel test test/extensions/filters/network/connection_limit:all` and didn't see my debug log.

Sounds good, then I will use `read_callbacks_->connection().dispatcher().createTimer()` to create my real logic timer.

Reply:
I'd recommend to read `bazel/README.md` first. But I built the binary with
```
bazel build //source/exe:envoy-static --define wasm=disabled
```
Then I run the built binary with
```
bazel-bin/source/exe/envoy-static -c envoy-config-network-conn-limit.yaml -l trace
```

And my `envoy-config-network-conn-limit.yaml` contains the following
```yaml
enable_dispatcher_stats: true
admin:
  access_log_path: /dev/null
  address:
    socket_address:
      address: 127.0.0.1
      port_value: 12000
static_resources:
  clusters:
    name: cluster_0
    connect_timeout: 0.25s
    circuit_breakers:
      thresholds:
      - priority: DEFAULT
        max_connections: 1000000000
        max_pending_requests: 1000000000
        max_requests: 1000000000
        max_retries: 1000000000
      - priority: HIGH
        max_connections: 1000000000
        max_pending_requests: 1000000000
        max_requests: 1000000000
        max_retries: 1000000000
    load_assignment:
      cluster_name: cluster_0
      endpoints:
        - lb_endpoints:
            - endpoint:
                address:
                  socket_address:
                    address: 127.0.0.1
                    port_value: 4500
  listeners:
    name: listener_0
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 10000
    filter_chains:
      filters:
      - name: connectionlimit
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.connection_limit.v3.ConnectionLimit
          stat_prefix: connection_limit_stats
          max_connections: 1
          delay: 0.2s
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: auto
          generate_request_id: false
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: backend
              domains:
              - "*"
              routes:
              - match:
                  prefix: "/"
                route:
                  cluster: cluster_0
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
              dynamic_stats: false
```

Reply:
Sure, I will read the `bazel/README.md`. Thanks very much, this is very helpful.
I will work on this and keep you posted.

Reply:
@rojkov Hi, I have updated my PR for adding delay before closing a connection. Please review, thanks!